### ISSUE TITLE
Scope of Fix

### ISSUE THREAD

**Issue by Krzysztof-Cieslak** [fsprojects/Fix#7]
Hey,

It was started on [twitter](https://twitter.com/k_cieslak/status/673129814757388288) but I think it's better to continue discussion here.

I would like to discuss scope of this project - from my point of view, I would rather focus on project / solution scaffolding which works very well with Paket and FAKE. Yeoman scaffolding theoretically has same purpose but it was designed for JS world (i.e. single projects, not multiple projects solutions) so I hit some walls while trying to add good support for things like Paket, FAKE and other "solution" level tools. Making own project which solves this problem is probably much easier than trying to hack some workarounds in yeoman... and Fix is already here and could be used as base for that.

Some my ideas (in no particular order):

1. Fix can be installed "globally" and added to PATH - installation could contains its own Paket and FAKE "installations" (updated on start of Fix or something like that). We could create `fix paket` and `fix fake` commands which would take normal Paket and FAKE arguments, invoke tool and pass those arguments - it would allow user to stop including paket.bootstraper.exe, paket.exe and dependency to FAKE in every solution.

2. Fix would be run always from solution folder, it would ask (or maybe some config file) for subfolder for project which user wants to create. It would allow us to easily make necessary changes in fsproj files (relative paths etc. )

3. Project templates would look similar to what we have right now for yeoman (Fix is using same templates as far as I know). But we would add paket.references file (we would need to update solution paket.dependency when adding new project to solution) and some kind of file defining specific FAKE target(s) required for this specific project which would be merged into solution build.fsx file. 

4. Fix would allow to add solution level tools (like FSharp.Formatting) and would also add specific FAKE targets for this tooling.

Summing up - IMHO, we need tool which would unify tooling stack that should be used with F# (outside of VS) - so project scaffolding, dependency manager (through Paket) and building (through FAKE).

cc: @forki @tpetricek @dsyme

**Comment by reidev275** [comment]
I think all of these are fantastic ideas.  It focuses on the original intent I had for Fix while extending it in a direction of need for the rest of the community.  

Fix uses the actual repo from the Project Templates, though there are a couple issues #1 and #2 which will probably need to be resolved relating to Paket.

**Comment by reidev275** [comment]
@Krzysztof-Cieslak ok so going back through the twitter conversation and your comments and wanted to dig into each of your points a bit more because I really like the ideas.

1.  So FIx would package versions of paket and fake which means we'd either want some sort of `fix paket update` command to update the paket version used by fix or be forced to deploy new versions of Fix when new versions of the other tools are released.  I think I prefer the `fix paket update` approach but could be swayed if a compelling reason exists to go a different direction.

2. It already does something very similar with `fix new [projectName]`.  That creates a new folder `./[projectName]` and places the chosen template within that folder.

3. I'm not very knowledgeable with paket so input would be very welcome there.

4. Same as 3.  Would like to know more but isn't something I've done.

**Comment by mastoj** [comment]
I just found this project, don't know how I missed it. It looks really awesome and . I have one question though regarding 1. @Krzysztof-Cieslak. If Fix comes with Paket and FAKE, how is the build supposed to work on a build server? Do you check in paket.exe and Fake.exe or do you install Fix on the build server? Are there any other options?

**Comment by mexx** [comment]
@mastoj If I got it right, Fix is for the developer to manage the project and solution files, the build server should never see it. About Paket and FAKE, only `Paket.Bootstrapper.exe` should be checked in. With it one can restore Paket and FAKE can be restored from Nuget via `paket.dependencies`.

**Comment by Krzysztof-Cieslak** [comment]
That's one option. Another would be installing Fix on build server and using it also there (`build.cmd` / `build.sh` file). "Deployment" process of Fix is not solved yet but it's pretty easy to download latest release from GitHub, unzip file and add directory to PATH on most build servers I've used.

Actually, I think that would be *my* preferable way of doing things.

**Comment by mastoj** [comment]
@mexx, @Krzysztof-Cieslak first wrote "it would allow user to stop including paket.bootstraper.exe, paket.exe and dependency to FAKE in every solution". 

My opinion is that it is better to include a small bootstraper then installing Fix on a build server. What if you're using appveyor or some other build service? In that case I think it really nice to have a simple build bootstraper that fetch what I need to start the build, in this case Paket.

**Comment by mexx** [comment]
I wouldn't go for the build script road, as it's already solved multiple times.
The support for project management on the developer site is not good and that would be what I would love to see addressed.
Btw. how does this compares to the `dotnet cli`?

**Comment by Krzysztof-Cieslak** [comment]
One use case doesn't exclude the other - if someone wants to use Fix just as development tool it works OK. If someone wants to integrate it into build process it will also work. It's not really solving the problem again - we are using Paket and FAKE, Fix just change  a bit how Paket and FAKE are distributed and invoked.

I have no idea at all about `dotnet cli` :)

**Comment by reidev275** [comment]
Good points @Krzysztof-Cieslak, I don't think the two are mutually exclusive.  The developer experience portion of the tool was my primary focus.  I think because FIx is now also a proxy for Fake and Paket it has the ability to be used in ways I never imagined which is a really cool development.  

@mexx I know very little about `dotnet cli`.  From their site it looks like they're probably solving some similar ground but from a C#, Nuget, msft best practice standpoint.

**Comment by mastoj** [comment]
@reidev275 @Krzysztof-Cieslak Would it make sense for Fix to use FAKE and Paket from subfolders if those are available and fallback to global versions if not? I mean if you're in a folder that contains a .paket folder it uses what ever is in that folder, and if you have a packages folder with FAKE you use that version. That should make it easier to be sure that you're doing the same thing on a developer machine and build server.

**Comment by Krzysztof-Cieslak** [comment]
@mastoj, take a look at #34.

And I guess it's time to close this. We have enough to discuss specific ideas in their own issues.
